### PR TITLE
ARROW-452: [C++/Python] Incorporate C++ and Python codebases for Feather file format

### PR DESCRIPTION
**Pull request by wesm**
The goal for this patch is to provide an eventual migration path for Feather (https://github.com/wesm/feather) users to use the batch and streaming Arrow file formats internally. Eventually the Feather metadata can be deprecated, but we will need to wait for the R community to build and ship Arrow bindings for R before that can happen. In the meantime, we won't need to maintain multiple Python/C++ codebases for the Python side of things. 

The test suite isn't yet passing because support for timestamps with time zones has not been implemented in the conversion to pandas.DataFrame, so I will do that when I can, but this can be reviewed in the meantime. 

I would upload a Gerrit code review, but there are some access control settings on gerrit.cloudera.org that need changing.

### CODE REVIEW

**Comment by wesm** [comment]
Another issue I have to deal with is that Feather uses int32 days since UNIX epoch for dates, and Arrow currently is int64 millis since the epoch, so zero copy date reads isn't possible at the moment

**Comment by wesm** [comment]
@xhochy I added a 32-bit date type. While it adds complexity having 2 different date types, since the int32 days-since-UNIX-epoch occurs quite frequently in the wild, not being able to load with zero copy would be a deficiency. Luckily adding new logical types doesn't involve too much new code

**Comment by xhochy** [comment]
@wesm If we want to keep this type, we should add a two JIRAs:

 * Add this type to the JVM implementation
 * Add a convenience cast function between the two types. (We should probably soon add a general cast interface, at least in `parquet-cpp` that would reduce some lines of code).

(I will probably have some time on Friday to review the changes here. Given the scale, Gerrit would be preferable but I'm ok with Github if it turns out to be problematic currently.)

**Comment by wesm** [comment]
@xhochy we need to turn on Forge Committer in Gerrit because it's rejecting the commits by authors that aren't registered in the system. I'll reach out to some folks at Cloudera to see if that's possible. 

I don't think we necessarily need to add the Date32 type to the Java implementation so long as we have a function to cast to the proper Date type in an IPC setting, we should certainly have a discussion about it though.

Agree about having casting functions

**Comment by wesm** [comment]
I should have the timestamp-with-tz issues resolved today, and I sent a message to ASF committers at Cloudera to change the settings on the Arrow project there so I can upload the review

**Comment by wesm** [comment]
@xhochy if it helps, I think I can divide this patch into two pieces, the first is the IPC refactor (that creates `arrow/loader.h`) and support machinery (i.e. the Date32Array etc.), and the 2nd part is the Feather code itself

**Comment by wesm** [comment]
@xhochy I split the unrelated changes into ARROW-605, so let's review that first and then I will rebase. going to get this branch passing now

**Comment by wesm** [comment]
Completing the timestamp+tz functionality is going to be a bit invasive because of pandas's "synthetic" dtypes, which means we need to create the Arrow type object before calling `PandasToArray`. After #364 and #365 are in I will take the opportunity to do some refactoring in pandas.cc to do this more cleanly

File: cpp/src/arrow/io/memory.cc
Comment (tebeka) [inline cpp/src/arrow/io/memory.cc]: Is this use somewhere?

**Comment by wesm** [comment]
Rebased. I will complete the datetime-with-tz issue when I can, maybe tomorrow or Saturday

**Review by xhochy — APPROVED** [review]
Everything looks good here 👍

**Comment by wesm** [comment]
I need to do some refactoring for the datetime+tz issue to get the test suite passing, so I'll ping you again when I have that all working

**Comment by wesm** [comment]
Actually, I'm going to disable those tests and merge this patch once the build is green. That way we can review the refactoring work more easily as a new patch

**Pull request closed without merging**
